File: src/lib/bodyweightdraft.test.ts
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import {
  INVALID_WEIGHT,
  parseBodyWeight,
  parseWeightAmount,
} from './bodyweightdraft'
import { NUMERIC_VALUE_MUST_BE_STRING } from './draft'
import { reservedTagError } from './tags'

const weightCases = JSON.parse(
  readFileSync(
    path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '../../testdata/weight-amount-cases.json',
    ),
    'utf8',
  ),
) as {
  invalidWeightError: string
  numericValueMustBeString: string
  accept: { input: string; stored: string }[]
  reject: { input: string; error: string }[]
}

describe('parseWeightAmount shared fixtures', () => {
  it('matches fixture error constants', () => {
    expect(INVALID_WEIGHT).toBe(weightCases.invalidWeightError)
    expect(NUMERIC_VALUE_MUST_BE_STRING).toBe(weightCases.numericValueMustBeString)
  })

  it.each(weightCases.accept)('accepts $input → $stored', ({ input, stored }) => {
    expect(parseWeightAmount(input)).toEqual({ ok: true, value: stored })
  })

  it.each(weightCases.reject)('rejects $input', ({ input, error }) => {
    expect(parseWeightAmount(input)).toEqual({ error })
  })

  it('rejects JSON number with numeric_value must be a decimal string', () => {
    expect(parseWeightAmount(75.5)).toEqual({
      error: NUMERIC_VALUE_MUST_BE_STRING,
    })
  })
})

describe('parseBodyWeight', () => {
  const base = {
    happened_at: '2026-08-02T08:00:00+08:00',
    numeric_value: '75.5',
    objective_context: 'morning weigh-in',
  }

  it('prepends body:weight and normalizes value', () => {
    const parsed = parseBodyWeight({
      ...base,
      subjective_interpretation: 'a bit heavy',
      tags: ['morning'],
    })
    expect('error' in parsed).toBe(false)
    if ('error' in parsed) return
    expect(parsed.numericValue).toBe('75.50')
    expect(parsed.tags).toEqual(['body:weight', 'morning'])
    expect(parsed.objectiveContext).toBe('morning weigh-in')
    expect(parsed.subjectiveInterpretation).toBe('a bit heavy')
  })

  it('allows omitted or empty tags (only body:weight)', () => {
    const a = parseBodyWeight(base)
    expect('error' in a).toBe(false)
    if ('error' in a) return
    expect(a.tags).toEqual(['body:weight'])

    const b = parseBodyWeight({ ...base, tags: [] })
    expect('error' in b).toBe(false)
    if ('error' in b) return
    expect(b.tags).toEqual(['body:weight'])
  })

  it('rejects reserved client tags', () => {
    expect(
      parseBodyWeight({ ...base, tags: ['body:weight'] }),
    ).toEqual({ error: reservedTagError('body:weight') })
    expect(
      parseBodyWeight({ ...base, tags: ['body:weight:x'] }),
    ).toEqual({ error: reservedTagError('body:weight:x') })
    expect(
      parseBodyWeight({ ...base, tags: ['transaction_entry'] }),
    ).toEqual({ error: reservedTagError('transaction_entry') })
  })

  it('rejects unknown keys', () => {
    expect(
      parseBodyWeight({ ...base, unit: 'kg' } as typeof base & { unit: string }),
    ).toEqual({ error: 'Unknown JSON key: unit' })
  })
})

File: faas/internal/bodyweightdraft/bodyweightdraft_test.go
package bodyweightdraft

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/mdk/digitaltwin2026/faas/internal/draft"
	"github.com/mdk/digitaltwin2026/faas/internal/tags"
)

type weightAmountAccept struct {
	Input  string `json:"input"`
	Stored string `json:"stored"`
}

type weightAmountReject struct {
	Input string `json:"input"`
	Error string `json:"error"`
}

type weightAmountCases struct {
	InvalidWeightError          string               `json:"invalidWeightError"`
	ErrNumericValueMustBeString string               `json:"numericValueMustBeString"`
	Accept                      []weightAmountAccept `json:"accept"`
	Reject                      []weightAmountReject `json:"reject"`
}

func loadWeightAmountCases(t *testing.T) weightAmountCases {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	root := filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
	b, err := os.ReadFile(filepath.Join(root, "testdata", "weight-amount-cases.json"))
	if err != nil {
		t.Fatalf("read weight-amount-cases: %v", err)
	}
	var cases weightAmountCases
	if err := json.Unmarshal(b, &cases); err != nil {
		t.Fatalf("parse weight-amount-cases: %v", err)
	}
	return cases
}

func TestParseWeightAmountSharedFixtures(t *testing.T) {
	cases := loadWeightAmountCases(t)
	if ErrInvalidWeight.Error() != cases.InvalidWeightError {
		t.Fatalf("ErrInvalidWeight constant drift: %q vs %q", ErrInvalidWeight, cases.InvalidWeightError)
	}
	if draft.ErrNumericValueMustBeString.Error() != cases.ErrNumericValueMustBeString {
		t.Fatalf("ErrNumericValueMustBeString drift")
	}
	for _, tc := range cases.Accept {
		got, err := ParseWeightAmount(tc.Input)
		if err != nil || got != tc.Stored {
			t.Fatalf("accept %q: got=%q err=%v want=%q", tc.Input, got, err, tc.Stored)
		}
	}
	for _, tc := range cases.Reject {
		_, err := ParseWeightAmount(tc.Input)
		if err == nil || err.Error() != tc.Error {
			t.Fatalf("reject %q: err=%v want=%q", tc.Input, err, tc.Error)
		}
	}
	_, err := ParseWeightAmount(float64(75.5))
	if err == nil || err.Error() != draft.ErrNumericValueMustBeString.Error() {
		t.Fatalf("JSON number: err=%v", err)
	}
}

func TestParseBodyWeight(t *testing.T) {
	raw := []byte(`{
		"happened_at": "2026-08-02T08:00:00+08:00",
		"numeric_value": "75.5",
		"objective_context": "morning weigh-in",
		"ai_analysis": "a bit heavy",
		"tags": ["morning"]
	}`)
	got, err := ParseBodyWeight(raw)
	if err != nil {
		t.Fatal(err)
	}
	if got.NumericValue != "75.50" {
		t.Fatalf("value=%q", got.NumericValue)
	}
	if len(got.Tags) != 2 || got.Tags[0] != tags.ReservedTagBodyWeight || got.Tags[1] != "morning" {
		t.Fatalf("tags=%v", got.Tags)
	}
}

func TestParseBodyWeightOmitsTags(t *testing.T) {
	raw := []byte(`{
		"happened_at": "2026-08-02T08:00:00+08:00",
		"numeric_value": "75",
		"objective_context": "x"
	}`)
	got, err := ParseBodyWeight(raw)
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Tags) != 1 || got.Tags[0] != tags.ReservedTagBodyWeight {
		t.Fatalf("tags=%v", got.Tags)
	}
	if got.NumericValue != "75.00" {
		t.Fatalf("value=%q", got.NumericValue)
	}
}

func TestParseBodyWeightRejectsReservedClientTag(t *testing.T) {
	raw := []byte(`{
		"happened_at": "2026-08-02T08:00:00+08:00",
		"numeric_value": "75",
		"objective_context": "x",
		"tags": ["body:weight"]
	}`)
	_, err := ParseBodyWeight(raw)
	want := tags.ReservedTagError("body:weight")
	if err == nil || err.Error() != want {
		t.Fatalf("err=%v want=%q", err, want)
	}
}

func TestParseBodyWeightRejectsDuplicateClientTag(t *testing.T) {
	raw := []byte(`{
		"happened_at": "2026-08-02T08:00:00+08:00",
		"numeric_value": "75.5",
		"objective_context": "x",
		"tags": ["morning", "morning"]
	}`)
	_, err := ParseBodyWeight(raw)
	if err == nil || err.Error() != `duplicate tag "morning"` {
		t.Fatalf("err: %v", err)
	}
}
